Updates to quest's signed distance example

kennyweiss opened the pull request:
# Summary

- This PR is a feature
- It adds the ability to run the signed distance query example in batched mode
- It also adds an option to `SignedDistance` and its C API to skip the computation of the sign, returning just the magnitude of the distance. This can be useful in applications that do not care about the sign
- It also converts the signed distance example's argument parsing to `CLI11`

rhornung67 commented on src/axom/quest/SignedDistance.hpp:
```diff
@@ -70,14 +70,16 @@ class SignedDistance
-   * \param [in] maxLevels max levels for spatial decomposition (optional).
+   * \param [in] maxLevels max levels for spatial decomposition.
+   * \param [in] computeSign indicates if distance queries should compute signs.
    * \note Default maxLevels is 5 if not specified.
```
maxlevels is required since it is not defaulted in the header declaration here. So this note is confusing.

kennyweiss replied:
Thanks @rhornung67 -- As you mentioned, there is no default, so this option is required.
I fixed this since the documentation was previously incorrect. 

(It might be confusing from the github view, but the change that you commented on was actually a removal of the word "optional").

gzagaris replied:
Thanks for fixing the comment. I think in some initial implementation `maxLevels` was indeed optional.  That was changed at some point but the documentation was not updated to reflect this change...

rhornung67 commented on RELEASE-NOTES.md:
```diff
@@ -10,6 +10,10 @@ The Axom project release numbers follow [Semantic Versioning](http://semver.org/
 ### Added
+- Added option to quest's `SignedDistance` class and C API to toggle whether
+  the distance query computes the sign.
+- Added batched option to accelerate quest's signed distance query example.
```
I think the statement about 'batched option' should be clarified --> what this means is that one can pass in a vector of point coordinates and process them in parallel, or something along those lines.

kennyweiss replied:
I'll try to rework this.

gzagaris commented on src/axom/quest/examples/quest_signed_distance_interface.cpp:
```diff
@@ -52,21 +55,74 @@ int numranks;
+      ->capture_default_str();
+
+    app.add_flag("--batched", this->use_batched_query, 
+      "uses a vectorized query instead of many serial queries")
```
I wouldn't call this `vectorized` -- just call it `parallel` or `parallel via OpenMP` to be more precise..

kennyweiss replied:
Technically, it is actually not necessarily `parallel` or `parallel via OpenMP` -- that depends on how axom is configured.
The only guarantee is that it is batched.

I'll try to rework this message.

gzagaris commented on src/axom/quest/examples/quest_signed_distance_interface.cpp:
```diff
@@ -52,21 +55,74 @@ int numranks;
+    minbb->needs(maxbb);
+    maxbb->needs(minbb);
+
+    app.add_flag("!--no-vtk", this->dump_vtk, "disables VTK output")
```
Does this need the `!` ?

kennyweiss replied:
The `!` tells `cli11` to invert the value when the flag is used.

I.e. If the user passes `--no-vtk` in as a command line argument, then the value of `this->dump_vtk` will be `false`.
Without the `!`, it would be `true`.

gzagaris replied:
Got it! Thanks for clarifying.

gzagaris commented on src/axom/quest/examples/quest_signed_distance_interface.cpp:
```diff
@@ -52,21 +55,74 @@ int numranks;
+    app.add_flag("!--no-vtk", this->dump_vtk, "disables VTK output")
+      ->capture_default_str();
+
+    app.add_flag("!--not-watertight", this->is_water_tight, "indicates that input is not water-tight")
```
Ditto. Does this need the `!` or is it a typo?

kennyweiss replied:
see my reply to the previous one.

gzagaris commented on src/axom/quest/examples/quest_signed_distance_interface.cpp:
```diff
@@ -52,21 +55,74 @@ int numranks;
+    app.get_formatter()->column_width(40);
+
+    // could throw an exception
+    app.parse( argc, argv);
```
How does `CLI` handle the `--help`, does it handle that internally?

kennyweiss replied:
Yes, It is automatic,

kennyweiss replied:
For convenience, here is the output:
```
>./examples/quest_signed_distance_interface_ex -h
Driver for signed distance query
Usage: ./examples/quest_signed_distance_interface_ex [OPTIONS]

Options:
  -h,--help                             Print this help message and exit
  -f,--file TEXT:FILE REQUIRED          specifies the input mesh file
  --dimension INT=3                     the problem dimension
  --maxLevels INT=15                    max levels of subdivision for the BVH
  --maxOccupancy INT=5                  max number of item per BVH bin
  --box-dims INT x 3                    the dimensions of the box mesh
  --box-min FLOAT x 3 Needs: --box-max  the lower corner of the box mesh
  --box-max FLOAT x 3 Needs: --box-min  the upper corner of the box mesh
  --no-vtk{false}                       disables VTK output
  --not-watertight{false}               indicates that input is not water-tight
  --use-shared                          stores the surface using MPI-3 shared memory
  --batched                             uses a single batched query on all points instead of many individual queries
  --ignore-signs                        distance query should ignore signs

```

gzagaris replied:
Nice!

gzagaris commented on src/axom/quest/SignedDistance.hpp:
```diff
@@ -70,14 +70,16 @@ class SignedDistance
                   int maxObjects,
-                  int maxLevels   );
+                  int maxLevels,
+                  bool computeSign );
```
Let's make `computeSign` optional here and always default to `true` since the class is called `SignedDistance`...

kennyweiss replied:
ok

kennyweiss commented on src/axom/quest/examples/quest_signed_distance_interface.cpp:
```diff
@@ -52,21 +55,74 @@ int numranks;
+  int maxLevels{15};
+  int maxOccupancy{5};
+  std::vector<int> box_dims{32,32,32};
+  std::vector<double> box_min;
```
This is for parsing into `CLI11`  It parses the input into the vectors and I've set that option to require three inputs.

gzagaris commented:
Also, please run shroud since there is a new API call that we should wrap to the Fortran API....

kennyweiss commented on src/axom/quest/examples/quest_signed_distance_interface.cpp:
```diff
@@ -52,21 +55,74 @@ int numranks;
+    app.add_option("--box-dims", box_dims, "the dimensions of the box mesh")
+      ->expected(3);
+
+    auto* minbb = app.add_option("--box-min", box_min, "the lower corner of the box mesh")
```
This does not return a pointer to the data. It returns a pointer to the `cli11` Option type (or something similar).  
The data actually gets parsed into the supplied ``std::vector`` `box_dims`.

We don't care about what the return type is here, we only need it so add a constraint a few lines down that if the user supplies `--box-min`, they also need to supply a ``--box-max``.

kennyweiss replied:
Note: I added a comment to the code.

kennyweiss commented on src/axom/quest/examples/quest_signed_distance_interface.cpp:
```diff
@@ -52,21 +55,74 @@ int numranks;
+  int ndims{3};
+  int maxLevels{15};
+  int maxOccupancy{5};
+  std::vector<int> box_dims{32,32,32};
```
We're not actually using that type in the code -- it is explicitly cast to an `axom::IndexType` below when it is passed into the `mint` mesh.

I think it is better to keep this as an `int` in case the user supplies a negative number, it is more likely to get caught below.

kennyweiss replied:
Nevermind -- `IndexType` is a signed int -- I'll make the change.

kennyweiss commented on src/axom/quest/interface/signed_distance.hpp:
```diff
@@ -175,6 +175,14 @@ void signed_distance_set_dimension( int dim );
+ *
+ * \note Options must be set before initializing the Signed Distance Query.
+ */
+void signed_distance_set_compute_distance( bool computeSign );
```
Thanks -- that is definitely what I intended!

kennyweiss commented on src/axom/quest/tests/quest_signed_distance.cpp:
```diff
@@ -131,7 +131,7 @@ TEST( quest_signed_distance, sphere_test )
   SLIC_INFO( "Generate BVHTree..." );
   // _quest_distance_cpp_init_start
-  axom::quest::SignedDistance< 3 > signed_distance( surface_mesh,true, 25, 10 );
+  axom::quest::SignedDistance< 3 > signed_distance( surface_mesh,true, 25, 10, true );
```
ok

kennyweiss commented on src/axom/quest/examples/quest_signed_distance_interface.cpp:
```diff
@@ -91,22 +147,36 @@ int main ( int argc, char** argv )
+  }
+  catch (const CLI::ParseError &e)
+  {
+    // TOOD convert to MPI parsing!
```
It is a change I have to make before merging.  

Since this app could be using MPI, I have to modify it so it cleans up MPI in case of a parsing error.

gzagaris commented on src/axom/quest/examples/quest_signed_distance_interface.cpp:
```diff
@@ -274,12 +276,16 @@ void generate_uniform_box_mesh( mint::UniformMesh*& mesh, Arguments& args)
+    
+    const primal::Point<int,3> bdims(args.box_dims.data(), 3);
+    SLIC_INFO( "constructing Uniform Mesh of resolution " << bdims );
+  }
```
Is this needed? 

You are constructing primal objects just so that they get printed?

kennyweiss replied:
Yes -- the output is significantly better from the primal objects.

Specifically, the Bounding box output provides the range of values.
E.g.:
```
{ min:  (0.03,  -2.1684e-19,-2.1684e-19); 
  max:  (0.0335,0.03, 0.03); 
  range:<0.0035,0.03, 0.03> }
```

This helped me (the user in this case) determine what bounding box I should provide to the code to get the desired results.

kennyweiss commented:
@rhornung67 @gzagaris  and others -- thanks for the thorough review.
I think I've addressed all the comments. Please review again at your convenience.

kennyweiss commented:
The `batched` option is significantly faster, as expected.
Here's what I got on a full `toss3` node (36 cores):
Without ``--batched``:
```
>./examples/quest_signed_distance_interface_ex -f ../data/quest/sphere.stl --box-dims 100 100 100 
[0][INFO]: initializing signed distance function...
[0][INFO]: input file: ../data/quest/sphere.stl
[0][INFO]: max_levels=15
[0][INFO]: max_occupancy=5
[0][INFO]: time to initialize: 0.00859259s
[0][INFO]: computing mesh bounds...
[0][INFO]: bounding box { min:(-5,-5,-5); max:(5,5,5); range:<10,10,10> }
[0][INFO]: constructing Uniform Mesh of resolution (100,100,100)
[0][INFO]: evaluating signed distance field on specified box mesh...
[0][INFO]: time to evaluate: 24.817s                                   # <--- HERE
[0][INFO]: writing vtk output
```

With ``--batched``:
```
>./examples/quest_signed_distance_interface_ex -f ../data/quest/sphere.stl --box-dims 100 100 100 --batched
[0][INFO]: initializing signed distance function...
[0][INFO]: input file: ../data/quest/sphere.stl
[0][INFO]: max_levels=15
[0][INFO]: max_occupancy=5
[0][INFO]: time to initialize: 0.00853686s
[0][INFO]: computing mesh bounds...
[0][INFO]: bounding box { min:(-5,-5,-5); max:(5,5,5); range:<10,10,10> }
[0][INFO]: constructing Uniform Mesh of resolution (100,100,100)
[0][INFO]: evaluating signed distance field on specified box mesh...
[0][INFO]: time to evaluate batched query: 0.983249s                   # <-- HERE
[0][INFO]: writing vtk output
```

agcapps commented on src/axom/quest/SignedDistance.hpp:
```diff
@@ -70,14 +70,16 @@ class SignedDistance
-   * \param [in] maxLevels max levels for spatial decomposition (optional).
+   * \param [in] maxLevels max levels for spatial decomposition.
+   * \param [in] computeSign indicates if distance queries should compute signs (optional).
    * \note Default maxLevels is 5 if not specified.
```
@kennyweiss , I see that the declaration of `SignedDistance()` provides a default value of `true` for argument `computeSign` (new line 82).  Thanks!

I also see the `\note` at (new) line 75 says "Default maxLevels is 5 if not specified."  I don't see this specified as a default in the function declaration.  Where is `maxLevels = 5` specified?

Please add the following line to the Doxygen after (new) line 75:  "\note Default computeSign is true if not specified."

Edit: I see that this comment is in the left column of the GitHub review UI... rather confusing.

kennyweiss replied:
Fixed.

gzagaris commented on src/axom/quest/interface/CMakeLists.txt:
```diff
@@ -13,6 +13,7 @@ set(YAML_INPUT_FILE quest_shroud.yaml)
 
 add_shroud(
     YAML_INPUT_FILE      ${YAML_INPUT_FILE}
+    YAML_OUTPUT_DIR      yaml
```
Is that new in `shroud`? What does it do exactly?

kennyweiss replied:
Shroud generates `yaml` files as part of its interface files.  This tells it where to place them.

When I ran shroud on quest, I noticed that two files were missing, and the yaml file was at the base directory.
(Note: There doesn't seem to be too much in the quest yaml file).

gzagaris replied:
Thanks for the explanation -- I wasn't aware of that parameter.

kennyweiss commented:
Passed all LC `host-configs`
